맛집 조회 및 상세 조회 API 구현 완료

SigLee2247 opened the pull request:
## What is this PR

- 맛집 조회 및 상세 조회 API 구현

## Changes



## Issue Number

feat #1

close #

SigLee2247 committed:
feat: 맛집 상세 조회 응답 dto 구현 #1
- 맛집 상세 조회 응답을 위한 DTO 구현
- Location 엔티티 정보 제공을 위한 LocationDto 구현

SigLee2247 committed:
feat: RestaurantService getDetails 메서드 구현 #1
- 구현을 위한 RestaurantRepository 구현
- 데이터 변환을 위한 RestaurantMapper 구현

SigLee2247 committed:
test: RestaurantService getDetails 테스트를 위한 Config 구현 #1
- 테스트 환경에서의 의존성 주입을 위한 Config 구현

SigLee2247 committed:
test: RestaurantService getDetails 단위 테스트 #1
- 테스트를 위한 스텁 데이터 생성
- 조회 성공 및 실패 테스트

SigLee2247 committed:
test: RestaurantController getDetails 테스트를 위한 스텁 구현 #1
- 시큐리티 적용으로 인한 시큐리티 stub 구현
- 시큐리티 TestConfig AuthStub 추가
- 테스트를 위한 Stub 추가

SigLee2247 committed:
feat: RestaurantController getDetails 구현 #1

SigLee2247 committed:
test: RestaurantController getDetails 구현 #1
- 맛집 조회 성공 테스트
- 맛집 조회 실패[비로그인] 테스트

SigLee2247 committed:
fit: Restaurant 상세 요청 시 데이터 추가 구현 #1
- Restaurant 위도 경도 추가
- 데이터 추가로 인한 테스트 코드 수정

SigLee2247 committed:
chore: 테스트 Lombok 사용을 위한 의존성 추가 #1

SigLee2247 committed:
update: getter Lombok 변환 #1

SigLee2247 committed:
update: 상세 조회 시 rating 데이터 추가 전달 #1
- 데이터 조회를 위한 양방향 매핑
- 데이터 한번에 조회를 위한 fetch join 구현

SigLee2247 committed:
feat: 키워드 별 데이터 조회를 위한 findPageByFilter 메서드 구현 #1
- QueryDSL을 활용한 동적 쿼리 작성

SigLee2247 committed:
feat: 페이징 데이터 전달을 위한 CustomPage 구현 #1

SigLee2247 committed:
fit: Restaurant 연관관계 수정 #1
- 기존의 location과의 연관관계 1:1 -> N:1 수정

SigLee2247 committed:
fit: Restaurant getRestaurant 메서드 구현을 위한 dto 구현 #1

SigLee2247 committed:
fit: Restaurant getRestaurant 메서드 구현을 위한 mapper 구현 #1

SigLee2247 committed:
fit: Restaurant getRestaurant 메서드 구현을 위한 service 구현 #1

SigLee2247 committed:
fit: Restaurant getRestaurant 메서드 구현을 위한 controller 구현 #1

SigLee2247 committed:
fit: Rating 오타 수정 #1
- Score -> score 오타 수정

jkde7721 commented on src/main/java/com/wanted/lunchmapservice/location/dto/LocationDto.java:
```diff
@@ -0,0 +1,11 @@
+
+import lombok.Builder;
+
+@Builder
+public record LocationDto(String cityName,
+                          String countryName,
+                          Double longitude,
+                          Double latitude) {
```
오 `record` 키워드는 처음 보는데 신기하네요👍
해당 DTO는 응답으로만 사용되는데 클래스 이름을 좀더 명시적으로 `LocationResponseDto`로 수정하시는 것은 어떤가요? + `RatingDto` → `RatingResponseDto`

ffolabear replied:
`LocationDto` 는 변경할일이 없기 때문에 `record` 로 선언하셨군요! 좋은 방법인것 같습니다.

SigLee2247 replied:
record를 처음 사용했는데 DTO와 같이 변경이 이루어지지 않는 코드의 경우는 조금 더 직관적으로 확인할 수 있는 것 같습니다.

jkde7721 commented on src/test/java/com/wanted/lunchmapservice/restaurant/controller/RestaurantControllerTest.java:
```diff
@@ -0,0 +1,107 @@
+    ObjectMapper objectMapper;
+    @Autowired
+    RestaurantMock mock;
+    @Autowired
+    AuthStub authStub;
+    @MockBean
+    RestaurantGetService getService;
+    @MockBean
+    UserDetailsServiceImpl detailsService;
+    String accessToken;
+    Principal principal;
+
+    @BeforeEach
+    void init() {
+        accessToken = authStub.createAccessToken();
+        principal = authStub.creatUserDetails();
+    }
```
`@WithMockUser` 어노테이션을 사용하면 따로 AuthStub을 지정해주지 않고도 테스트가 가능한데, 혹시 AuthStub을 사용하신 이유가 따로 있을까요?

SigLee2247 replied:
해당 부분은 통합테스트를 진행하려고 했던 코드를 Controller 단위 테스트로 변경하는 과정에서 수정하지 못했던 것 같습니다.
해당 부분 수정했습니다

jkde7721 commented on src/main/java/com/wanted/lunchmapservice/restaurant/service/RestaurantGetService.java:
```diff
@@ -0,0 +1,41 @@
+        return mapper.toResponseDto(result);
+    }
+
+    public ResponseDto<CustomPage<GetRestaurantSimpleResponseDto>> getRestaurant(@ModelAttribute RestaurantGetFilterDto request, Pageable pageable) {
+        return mapper.toResponseDto(repository.findPageByFilter(pageable,request));
+    }
+
+    private Restaurant validRestaurant(Long restaurantId) {
+        return repository.findById(restaurantId)
+                .orElseThrow(() -> new CommonException(HttpStatus.CONFLICT, "맛집 정보가 존재하지 않습니다."));
+    }
```
`getRestaurant` 메소드의 파라미터에서 `@ModelAttribute` 어노테이션 실수로 작성하신 것 같아요.
`validRestaurant` 메소드에서 맛집 정보가 존재하지 않을 때에는 NOT_FOUND 상태 코드가 더 적절하지 않을까 생각하는데 혹시 다른 이유가 있을까요?

SigLee2247 replied:
1. Service Code에 @ModelAttribute 이 붙어 있는 것은 오타입니다.
2. findById의 예외를 처리할 떄 상태 코드를 Conflict로 했던 것은 작업하면서 있었던 실수 였던 것 같습니다.
- 해당 부분에 대한 수정은 진행했고, 실수가 발생하지 않도록 노력하겠습니다

jkde7721 commented on src/main/java/com/wanted/lunchmapservice/restaurant/repository/RestaurantRepository.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.data.repository.query.Param;
+
+public interface RestaurantRepository extends JpaRepository<Restaurant,Long> , RestaurantQueryRepository{
+    @Query("SELECT r FROM Restaurant r " +
+            "LEFT JOIN FETCH r.ratingList rr " +
+            "LEFT JOIN FETCH rr.user " +
+            "WHERE r.id = :id")
+    Optional<Restaurant> findById(@Param("id") Long id);
```
`findById` 메소드가 `Restaurant` 엔티티만을 조회해오는 것이 아닌 다른 엔티티들도 페치조인해서 가져오므로, `findByIdFetch` 등으로 메소드 이름 변경하는 것은 어떤가요? 
또한 이 경우 만약 다른 분이 `findById` 메소드를 이미 호출해서 사용하고 있다면 예상치 않게 동작할 가능성도 있을 것 같습니다.

ffolabear replied:
`findByIdFetch` 라고 메소드 네이밍시 JPA 에서 에러가 발생하는 일은 혹시 없을까요? 최근에 JPA 메소드 네이밍에서 에러가 많이 발생했어서 여쭤봅니다 🥲

SigLee2247 replied:
`findByIdFetch`를 작성하고 위에서 `@Query`를 통해 SQL문을 제공하기 때문에 문제가 없을 것이라고 생각합니다
`findById` 메서드 `findByIdFatch` 메서드로 수정 완료 했습니다.

jkde7721 reviewed (commented):
구현하시느라 수고많으셨습니다!

ffolabear commented on src/main/java/com/wanted/lunchmapservice/restaurant/dto/GetRestaurantDetailResponseDto.java:
```diff
@@ -0,0 +1,15 @@
+                                             Double averageScore,
+                                             List<RatingDto> ratingList) {
+
+}
```
저는 이런 요청을 처리할때는 `Request~` 라는 이름으로, 응답할때는 `Response~` 를 붙여셔 네이밍 했습니다. Http 메소드로 표현하기 보다는 요청 / 응답을 표현하는 것이 더 중요하다고 생각했는데, 기호님 생각은 어떠신가요?

SigLee2247 replied:
해당 부분 말씀하신 이름 방식으로 변경했습니다.
- 지금 작성하는 방식과 다른 방식 모두 이상이 없을 것이라 판단해 제안하셨던 방식으로 변경했습니다

ffolabear commented on src/test/java/com/wanted/lunchmapservice/restaurant/service/RestaurantGetServiceTest.java:
```diff
@@ -0,0 +1,72 @@
+                .isInstanceOf(CommonException.class)
+                .hasMessage("맛집 정보가 존재하지 않습니다.");
+    }
+}
```
꼼꼼한 테스트코드 작성에서 많이 배우고 갑니다!

ffolabear reviewed (commented):
고생하셨습니다!

SigLee2247 committed:
refactor: 메서드 이름 수정 #1
- findById -> findByIdFetch

SigLee2247 committed:
refactor: 구현 메서드 Override 명시 #1

SigLee2247 committed:
test: ControllerTest 시 Mockito 를 활용한 로그인 #1
- WithMokUser 어노테이션을 활용한 인증 정보 인증

SigLee2247 committed:
fit: RestaurantGetService 오타 수정 #1
- 불필요한 어노테이션 제거

SigLee2247 committed:
fit: RestaurantGetService 조회 실패 시 HttpCode 변경 #1
- Conflict -> Not Found

SigLee2247 committed:
rename: Dto 이름 수정 #1
- Request / Response + HttpMethod + DomainName + Dto 의 형식으로 DTO 이름 수정
- 수정에 따라 파라미터 네임 수정
